### PR TITLE
core: implement Notary contract

### LINKED ISSUE #1545: P2P notary contract and attribute
Moving on to https://github.com/neo-project/neo/issues/1573#issuecomment-704874472 implementation. This issue is about three parts of this problem:
 * "P2P Notary" designated role
   Quite simple.
 * "Notary assisted" transaction attribute type
   Let's use `FEE` of 0.1 GAS for it. At the moment we have 0.5-1.5 GAS system fee overhead for signature collection per signer (plus 0.01 for network fee), which gives us overall 3.5-10.5 GAS overhead per multisign transaction. 0.1 sets the overall overhead of this service to 0.8 GAS in a typical "5 out of 7" scenario which seems to be economically reasonable.
 * notaries native contract
   The deposit scheme can be changed following NEP-17 `onPayment` change, but we still need to pass `till` parameter, so the default `onPayment` behavior could be to set some default number of blocks (5760?) and then a separate call would allow to update this value. The contract must handle `verify` (1 out of N multisig) and `onPersist` (distributing GAS to notaries) logic correctly.

All of this logic should be dependent on `P2PSigExtensions` setting.

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #1545

### CODE REVIEW

File: pkg/core/native/notary.go [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+// Notary represents Notary native contract.
+type Notary struct {
+	interop.ContractMD
+	GAS   *GAS
+	Desig *Designate
+}
+
+const (
+	notaryName       = "Notary"
+	notaryContractID = -6
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:30]: `-100500`. It's better for us to minimize chance of clashing with future official Neo contracts.

File: pkg/core/native/notary.go [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+
+	// prefixDeposit is a prefix for storing Notary deposits.
+	prefixDeposit = 1
+	// notaryRewardFactor is a factor by which notary node reward per notaries is multiplied
+	// to make calculations more precise.
+	notaryRewardFactor = 100_000_000
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:36]: I'm not sure it improves anything, we only do simple division of `FEE` by `N`.

File: pkg/core/native/notary.go [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+	// defaultDepositLockedHeight is a default chain's height until which deposit is locked.
+	defaultDepositLockedHeight uint32 = 5760 // TODO
+)
+
+// newNotary returns Notary native contract.
+func newNotary() *Notary {
+	n := &Notary{ContractMD: *interop.NewContractMD(notaryName)}
+	n.ContractID = notaryContractID
+
+	desc := newDescriptor("onPayment", smartcontract.VoidType,
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:46]: We desperately need NEP-17 changes...

File: pkg/core/native/notary.go [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+		manifest.NewParameter("from", smartcontract.Hash160Type),
+		manifest.NewParameter("amount", smartcontract.IntegerType))
+	md := newMethodAndPrice(n.onPayment, 100_0000, smartcontract.AllowStates)
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:49]: `AllowModifyStates`

File: pkg/core/native/notary.go [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+	n.AddMethod(md, desc, true)
+
+	desc = newDescriptor("lockDepositUntil", smartcontract.BoolType,
+		manifest.NewParameter("addr", smartcontract.Hash160Type),
+		manifest.NewParameter("till", smartcontract.IntegerType))
+	md = newMethodAndPrice(n.lockDepositUntil, 100_0000, smartcontract.AllowStates)
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:55]: `AllowModifyStates`

File: pkg/core/native/notary.go [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+	n.AddMethod(md, desc, true)
+
+	desc = newDescriptor("withdraw", smartcontract.BoolType,
+		manifest.NewParameter("from", smartcontract.Hash160Type),
+		manifest.NewParameter("to", smartcontract.Hash160Type))
+	md = newMethodAndPrice(n.withdraw, 100_0000, smartcontract.AllowStates)
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:61]: `AllowModifyStates`

File: pkg/core/native/notary.go [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+	md = newMethodAndPrice(n.balanceOf, 100_0000, smartcontract.AllowStates)
+	n.AddMethod(md, desc, true)
+
+	desc = newDescriptor("expirationOf", smartcontract.IntegerType,
+		manifest.NewParameter("addr", smartcontract.Hash160Type))
+	md = newMethodAndPrice(n.expirationOf, 100_0000, smartcontract.AllowStates)
+	n.AddMethod(md, desc, true)
+
+	desc = newDescriptor("verify", smartcontract.BoolType)
+	md = newMethodAndPrice(n.verify, 100_0000, smartcontract.NoneFlag)
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:75]: `AllowStates`?

File: pkg/core/native/notary.go [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+		}
+	}
+	if nFees == 0 {
+		return nil
+	}
+	notaries, err := n.GetNotaryNodes(ic.DAO)
+	if err != nil {
+		return fmt.Errorf("failed to get notary nodes: %w", err)
+	}
+	singleReword := calculateNotaryReword(nFees, len(notaries))
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:119]: reward

File: pkg/core/native/notary.go [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+// till the specified chain's height.
+func (n *Notary) onPayment(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	from := toUint160(args[0])
+	amount := toBigInt(args[1])
+
+	till := defaultDepositLockedHeight
+	deposit := n.getDepositFor(ic.DAO, from)
+	if deposit == nil {
+		if amount.Cmp(big.NewInt(2*transaction.NotaryServiceFeeFactor)) < 0 {
+			return stackitem.Null{} // TODO: call ABORT instead
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:136]: I guess we need some `(*VM).Abort`, just panicing here is not enough as it'll be treated as an exception that could be handled.

File: pkg/core/native/notary.go [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+		}
+		deposit = &state.Deposit{Amount: new(big.Int)}
+	} else {
+		till = deposit.Till
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:140]: Maybe extend it by `defaultDepositLockedHeight` if `Till` is obsolete already.

File: pkg/core/native/notary.go [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+	}
+	return nil
+}
+
+// onPayment records deposited amount as belonging to "from" address with a lock
+// till the specified chain's height.
+func (n *Notary) onPayment(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	from := toUint160(args[0])
+	amount := toBigInt(args[1])
+
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:131]: You need to check the calling script hash here first of all, it should be GAS (otherwise abort the execution).

File: pkg/core/native/notary.go [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+	}
+	till := uint32(toBigInt(args[1]).Int64())
+	if till < ic.Chain.BlockHeight() {
+		return stackitem.NewBool(false)
+	}
+	deposit := n.getDepositFor(ic.DAO, addr)
+	if deposit == nil {
+		return stackitem.NewBool(false)
+	}
+	if till < deposit.Till { // TODO: do we really need this?
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:168]: There can be some transactions in-flight and this can allow withdrawing the deposit before they settle.

File: pkg/core/native/notary.go [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+		panic(fmt.Errorf("failed to put deposit for %s into the storage: %w", to.StringBE(), err))
+	}
+	return stackitem.NewBool(true)
+}
+*/
+
+// withdraw sends all deposited GAS for "from" address to "to" address.
+func (n *Notary) withdraw(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	from := toUint160(args[0])
+	to := toUint160(args[1])
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:219]: I don't think `toUint160` handles NULL correctly for this case.

File: pkg/core/native/notary.go [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+		to = from
+	}
+	deposit := n.getDepositFor(ic.DAO, from)
+	if deposit == nil {
+		return stackitem.NewBool(false)
+	}
+	if ic.Chain.BlockHeight() < deposit.Till {
+		return stackitem.NewBool(false)
+	}
+	err := n.GAS.TransferInternal(ic, from, to, deposit.Amount)
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:230]: Not `from`, but from contract's address.

File: pkg/core/native/notary.go [hunk 15]
```diff
@@ -0,0 +1,345 @@
...
+}
+
+// expirationOf Returns deposit lock height for specified address.
+func (n *Notary) expirationOf(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	acc := toUint160(args[0])
+	deposit := n.getDepositFor(ic.DAO, acc)
+	if deposit == nil {
+		panic(fmt.Errorf("deposit for %s not found", acc.StringBE()))
+	}
+	return stackitem.NewBigInteger(big.NewInt(int64(deposit.Till)))
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:257]: `stackitem.Make()`?

File: pkg/core/native/notary.go [hunk 16]
```diff
@@ -0,0 +1,345 @@
...
+	deposit.Till = till
+	if err := n.putDepositFor(ic.DAO, deposit, to); err != nil {
+		panic(fmt.Errorf("failed to put deposit for %s into the storage: %w", to.StringBE(), err))
+	}
+	return stackitem.NewBool(true)
+}
+*/
+
+// withdraw sends all deposited GAS for "from" address to "to" address.
+func (n *Notary) withdraw(ic *interop.Context, args []stackitem.Item) stackitem.Item {
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:217]: Witness check is missing here.

File: pkg/core/native/notary.go [hunk 17]
```diff
@@ -0,0 +1,345 @@
...
+	deposit := n.getDepositFor(ic.DAO, acc)
+	if deposit == nil {
+		panic(fmt.Errorf("deposit for %s not found", acc.StringBE()))
+	}
+	return stackitem.NewBigInteger(big.NewInt(int64(deposit.Till)))
+}
+
+// verify checks whether the transaction was signed by one of the notaries.
+func (n *Notary) verify(ic *interop.Context, _ []stackitem.Item) stackitem.Item {
+	sig, err := ic.VM.Estack().Pop().Array()[0].TryBytes()
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:262]: Can we just specify that this method has a parameter? And I'm not sure why there is an `Array` here.

File: pkg/core/native/notary.go [hunk 18]
```diff
@@ -0,0 +1,345 @@
...
+					return stackitem.NewBool(false)
+				}
+				needFee = big.NewInt(tx.NetworkFee + tx.SystemFee)
+			}
+		} else if needFee != nil {
+			balance := ic.Chain.GetUtilityTokenBalance(signer.Account)
+			if balance != nil && balance.Cmp(needFee) >= 0 {
+				hasEnoughFunds = true
+			}
+
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:288]: Blank line.

File: pkg/core/native/notary.go [hunk 19]
```diff
@@ -0,0 +1,345 @@
...
+		signedByNotary, hasEnoughFunds bool
+		needFee                        *big.Int
+	)
+	for _, signer := range tx.Signers {
+		if signer.Account == n.Hash {
+			signedByNotary = true
+			if signer.Scopes != transaction.None {
+				return stackitem.NewBool(false)
+			}
+			if signer.Account.Equals(tx.Sender()) {
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:277]: `if i == 0` effectively.

File: pkg/core/native/notary.go [hunk 20]
```diff
@@ -0,0 +1,345 @@
...
+				if len(tx.Signers) == 1 {
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:278]: Maybe even `!= 2`, we need one cosigner that pays the bills actually.

File: pkg/core/native/notary.go [hunk 21]
```diff
@@ -0,0 +1,345 @@
...
+					return stackitem.NewBool(false)
+				}
+				needFee = big.NewInt(tx.NetworkFee + tx.SystemFee)
+			}
+		} else if needFee != nil {
+			balance := ic.Chain.GetUtilityTokenBalance(signer.Account)
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:284]: `getDepositFor`

File: pkg/core/native/notary.go [hunk 22]
```diff
@@ -0,0 +1,345 @@
...
+			if balance != nil && balance.Cmp(needFee) >= 0 {
+				hasEnoughFunds = true
+			}
+
+		}
+	}
+	if needFee != nil && !hasEnoughFunds {
+		return stackitem.NewBool(false)
+	}
+	if signedByNotary && len(tx.GetAttributes(transaction.NotaryAssistedT)) == 0 {
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:294]: Every transaction here MUST be signed by notary. And every transaction MUST have this attribute.

File: pkg/core/native/notary.go [hunk 23]
```diff
@@ -0,0 +1,345 @@
...
+	return &n.ContractMD
+}
+
+// Initialize initializes Notary native contract and implements Contract interface.
+func (n *Notary) Initialize(ic *interop.Context) error {
+	return nil
+}
+
+// OnPersist implements Contract interface.
+func (n *Notary) OnPersist(ic *interop.Context) error {
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:104]: I've just noticed that we're actually missing one important piece in the original description of `OnPersist`, we must burn the deposit made for transactions where the sender is notary contract (`tx.NetworkFee + tx.SystemFee` amount, this will be burned by GAS contract from notary contract address, so we need to reflect that in our deposits).

File: pkg/core/native/notary.go [hunk 24]
```diff
@@ -0,0 +1,345 @@
...
+
+	notaries, err := n.GetNotaryNodes(ic.DAO)
+	if err != nil {
+		panic(fmt.Errorf("failed to get notary nodes: %w", err))
+	}
+	hash := tx.GetSignedHash().BytesBE()
+	var verified bool
+	for _, n := range notaries {
+		if n.Verify(sig, hash) {
+			verified = true
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:306]: break

**Review by roman-khimov — CHANGES_REQUESTED** [review]
Looks good though. The contract itself is not that complicated.

File: pkg/core/native/notary.go (hunk 9 above)
Comment (AnnaShaleva) [reply pkg/core/native/notary.go]: Added panic.

File: pkg/core/native/notary.go (hunk 10 above)
Comment (AnnaShaleva) [reply pkg/core/native/notary.go]: Fixed.

File: pkg/core/native/notary.go (hunk 13 above)
Comment (AnnaShaleva) [reply pkg/core/native/notary.go]: Fixed, now Null is allowed.

File: pkg/core/native/notary.go (hunk 14 above)
Comment (AnnaShaleva) [reply pkg/core/native/notary.go]: That was done occasionally, fixed.

File: pkg/core/native/notary.go (hunk 17 above)
Comment (AnnaShaleva) [reply pkg/core/native/notary.go]: I thought that arguments are not popped from the stack by this moment for `verify`. But then I looked carefully at the code, so in fact, they are. Fixed.

File: pkg/core/native/notary.go (hunk 20 above)
Comment (AnnaShaleva) [reply pkg/core/native/notary.go]: Fixed. Do the users of this service know that there should be not more than 2 signers?

File: pkg/core/blockchain_test.go [hunk 25]
```diff
@@ -728,7 +728,7 @@ func TestVerifyTx(t *testing.T) {
 			})
 			t.Run("Enabled, positive", func(t *testing.T) {
 				bc.config.P2PSigExtensions = true
-				tx := getNotaryAssistedTx(1, (1+1)*transaction.NotaryServiceFeeFactor)
+				tx := getNotaryAssistedTx(1, (1+1)*transaction.NotaryServiceFeePerKey)
```
Comment (roman-khimov) [inline pkg/core/blockchain_test.go:731]: It belongs to different commit.

File: pkg/core/native/notary.go [hunk 26]
```diff
@@ -0,0 +1,347 @@
...
+						return fmt.Errorf("failed to update deposit for %s: %w", payer.Account.StringBE(), err)
+					}
+				}
+			}
+		}
+	}
+	if nFees == 0 {
+		return nil
+	}
+	singleReword := calculateNotaryReward(nFees, len(notaries))
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:139]: singleReward

**Review by roman-khimov — CHANGES_REQUESTED** [review]
Good. More tests are still needed, though. @fyrchik, please take a look also, this PR is important one.

File: pkg/core/native/notary.go (hunk 3 above)
Comment (roman-khimov) [reply pkg/core/native/notary.go:43]: I think we'll settle NEP-17 before this, so it should be fine.

File: pkg/core/native/notary.go (hunk 20 above)
Comment (roman-khimov) [reply pkg/core/native/notary.go]: > fallback tx
This transaction has P2P Notary contract as a sender and service request sender as an additional signer.

Already in neo-project/neo#1573 and this is gonna be something to check for in #1547 (#1546 can be little more permissive to simplify things for non-notary nodes).

File: pkg/core/native/notary.go [hunk 27]
```diff
@@ -0,0 +1,347 @@
...
+		return stackitem.NewBool(false)
+	}
+	var (
+		signedByNotary, hasEnoughFunds bool
+		needFee                        *big.Int
+	)
+	for i, signer := range tx.Signers {
+		if signer.Account == n.Hash {
+			signedByNotary = true
+			if signer.Scopes != transaction.None {
```
Comment (fyrchik) [inline pkg/core/native/notary.go:273]: This is checked during `Blockchain.verifyTxAttributes`, no need to check here.

File: pkg/core/native/notary.go [hunk 28]
```diff
@@ -0,0 +1,347 @@
...
+				return stackitem.NewBool(false)
+			}
+			if i == 0 {
+				if len(tx.Signers) != 2 {
```
Comment (fyrchik) [inline pkg/core/native/notary.go:277]: I think, this check can be moved to `verifyTxAttributes`.

File: pkg/core/native/notary.go [hunk 29]
```diff
@@ -0,0 +1,347 @@
...
+				needFee = big.NewInt(tx.NetworkFee + tx.SystemFee)
+			}
+		} else if needFee != nil {
+			balance := n.getDepositFor(ic.DAO, signer.Account)
+			if balance != nil && balance.Amount.Cmp(needFee) >= 0 {
+				hasEnoughFunds = true
+			}
+		}
+	}
+	if !signedByNotary {
```
Comment (fyrchik) [inline pkg/core/native/notary.go:289]: Can this be true? At the start we check presence of attribute and attribute verification requires having contract as signer.

File: pkg/core/native/notary.go [hunk 30]
```diff
@@ -0,0 +1,347 @@
...
+				return stackitem.NewBool(false)
+			}
+			if i == 0 {
+				if len(tx.Signers) != 2 {
+					return stackitem.NewBool(false)
+				}
+				needFee = big.NewInt(tx.NetworkFee + tx.SystemFee)
+			}
+		} else if needFee != nil {
+			balance := n.getDepositFor(ic.DAO, signer.Account)
```
Comment (fyrchik) [inline pkg/core/native/notary.go:283]: `needFee != nil` only if `Notary` is the sender of the transaction. Maybe we can simplify all this loop to checking if `tx.Sender() == n.Hash` and second signer has sufficient deposit?

**Review by fyrchik — DISMISSED** [review]

**Comment by fyrchik** [comment]
Also, there are typos in commit messages afab4cda and 807a4bd.

File: pkg/core/native/contract.go [hunk 31]
```diff
@@ -49,7 +49,7 @@ func (cs *Contracts) ByName(name string) interop.Contract {
 
 // NewContracts returns new set of native contracts with new GAS, NEO and Policy
 // contracts.
-func NewContracts() *Contracts {
+func NewContracts(p2pSigExtensionsEnabled bool) *Contracts {
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/core/native/contract.go:53]: Probably not yet worth it, too much useless data there.

File: pkg/core/native/notary.go [hunk 32]
```diff
@@ -0,0 +1,347 @@
...
+// Notary represents Notary native contract.
+type Notary struct {
+	interop.ContractMD
+	GAS   *GAS
+	Desig *Designate
+}
+
+const (
+	notaryName       = "Notary"
+	notaryContractID = -100500
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/core/native/notary.go:30]: We're trying to avoid clashing with future official Neo contracts for now. If and when it's to be accepted it can be changed.

File: pkg/core/native/notary.go [hunk 33]
```diff
@@ -0,0 +1,347 @@
...
+
+	// prefixDeposit is a prefix for storing Notary deposits.
+	prefixDeposit = 1
+	// depositLockedHeightIncrement is the number of blocks denoting the duration of deposit lock.
+	depositLockedHeightIncrement uint32 = 5760
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/core/native/notary.go:35]: This is a very different constant.

File: pkg/core/native/notary.go [hunk 34]
```diff
@@ -0,0 +1,347 @@
...
+		if tx.HasAttribute(transaction.NotaryAssistedT) {
+			if notaries == nil {
+				notaries, err = n.GetNotaryNodes(ic.DAO)
+				if err != nil {
+					return fmt.Errorf("failed to get notary nodes: %w", err)
+				}
+			}
+			nKeys := tx.GetAttributes(transaction.NotaryAssistedT)[0].Value.(*transaction.NotaryAssisted).NKeys
+			nFees += int64(nKeys) + 1
+			if tx.Sender() == n.Hash {
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/core/native/notary.go:118]: No, it won't. Some (actually, most of expected transactions) will have `n.Hash` in cosigners.

File: pkg/core/transaction/not_valid_before.go [hunk 35]
```diff
@@ -14,6 +15,13 @@ type NotValidBefore struct {
 // DecodeBinary implements io.Serializable interface.
 func (n *NotValidBefore) DecodeBinary(br *io.BinReader) {
 	bytes := br.ReadVarBytes(4)
+	if br.Err != nil {
+		return
+	}
+	if len(bytes) != 4 {
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/core/transaction/not_valid_before.go:21]: You can read 0-4 bytes above.

File: pkg/core/transaction/notary_assisted.go [hunk 36]
```diff
@@ -0,0 +1,37 @@
...
+	NKeys uint8 `json:"nkeys"`
+}
+
+// DecodeBinary implements io.Serializable interface.
+func (n *NotaryAssisted) DecodeBinary(br *io.BinReader) {
+	bytes := br.ReadVarBytes(1)
+	if br.Err != nil {
+		return
+	}
+	if len(bytes) != 1 {
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/core/transaction/notary_assisted.go:23]: You can read 0 bytes above.

File: pkg/core/native/notary.go (hunk 32 above)
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/core/native/notary.go]: Implemented
```
const reservedContractID = -100
const notaryContractID = reservedContractID-1
```

File: pkg/core/native/notary.go (hunk 29 above)
Comment (AnnaShaleva) [reply pkg/core/native/notary.go]: I refactored this code, it's simpler now.

**Comment by AnnaShaleva** [comment]
Finally, tests are done. @roman-khimov, I removed `lockDepositUntil` native Notary method, because `onPayment` can handle empty payments to update `till`. 

There's one more TODO left, what should we do in case if provided `till` is invalid? We can either (1) set the default value for `till` or (2) return an error. The second option is implemented at the current moment.

**Comment by roman-khimov** [comment]
> because onPayment can handle empty payments to update till

I don't think it's very convenient to the user. Suppose I have 10 GAS deposited already and I need to renew it. Requiring to send more GAS is fishy and sending 0 GAS just costs more than it should (going through two contracts instead of one).

> There's one more TODO left, what should we do in case if provided till is invalid? We can either (1) set the default value for till or (2) return an error.

Given that `data` is a part of the `transfer` now, I think we can just fail, so (2), to use notaries you need to supply this data, period.

File: pkg/core/helper_test.go [hunk 37]
```diff
@@ -411,3 +415,41 @@ func addNetworkFee(bc *Blockchain, tx *transaction.Transaction, sender *wallet.A
 	tx.NetworkFee += int64(size) * bc.FeePerByte()
 	return nil
 }
+
+func invokeNativeContractMethod(chain *Blockchain, hash util.Uint160, method string, args ...interface{}) (*state.AppExecResult, error) {
```
Comment (roman-khimov) [inline pkg/core/helper_test.go:419]: I guess a lot of code could make use of it.

File: pkg/core/blockchain.go [hunk 38]
```diff
@@ -1388,6 +1388,21 @@ func (bc *Blockchain) verifyTxAttributes(tx *transaction.Transaction) error {
...
+						return fmt.Errorf("%w: Notary contract signer is allowed to have None scope only", ErrInvalidAttribute)
+					}
+					signedByNotary = true
+				}
+			}
+			if !signedByNotary {
+				return fmt.Errorf("%w: NotaryAssisted attribute was found, but transaction is not signed by the Notary native contract", ErrInvalidAttribute)
+			}
+			if tx.Sender() == bc.contracts.Notary.Hash && len(tx.Signers) != 2 {
+				return fmt.Errorf("%w: Notary contract can not pay the fee", ErrInvalidAttribute)
```
Comment (roman-khimov) [inline pkg/core/blockchain.go:1404]: Is this case tested?

File: pkg/core/native/notary.go [hunk 39]
```diff
@@ -0,0 +1,358 @@
...
+	sig, err := args[0].TryBytes()
+	if err != nil {
+		panic(fmt.Errorf("failed to get signature bytes: %w", err))
+	}
+	tx := ic.Tx
+	if len(tx.GetAttributes(transaction.NotaryAssistedT)) == 0 {
+		return stackitem.NewBool(false)
+	}
+	if tx.Sender() == n.Hash {
+		payer := tx.Signers[1].Account
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:301]: This will fault easily now `if len(tx.Signers) == 0`, because attributes are checked after witnesses.

Comment (roman-khimov) [reply pkg/core/native/notary.go:301]: And maybe we better leave only witness presence check in attribute check and move the rest of this logic into `verify`.

File: pkg/core/native/notary.go [hunk 40]
```diff
@@ -0,0 +1,358 @@
...
+		manifest.NewParameter("addr", smartcontract.Hash160Type))
+	md = newMethodAndPrice(n.expirationOf, 100_0000, smartcontract.AllowStates)
+	n.AddMethod(md, desc, true)
+
+	desc = newDescriptor("verify", smartcontract.BoolType,
+		manifest.NewParameter("signature", smartcontract.SignatureType))
+	md = newMethodAndPrice(n.verify, 100_0000, smartcontract.AllowStates)
+	n.AddMethod(md, desc, false)
+
+	desc = newDescriptor("name", smartcontract.StringType)
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:78]: We no longer need it, it's now a part of the manifest.

File: pkg/core/native/notary.go [hunk 41]
```diff
@@ -0,0 +1,358 @@
...
+	}
+	from := toUint160(args[0])
+	to := from
+	amount := toBigInt(args[1])
+	data, ok := args[2].(*stackitem.Array)
+	if !ok || len(data.Value().([]stackitem.Item)) != 2 {
+		panic(errors.New("invalid data parameter"))
+	}
+	additionalParams := data.Value().([]stackitem.Item)
+	if bytes, err := additionalParams[0].TryBytes(); err == nil {
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:162]: It is either Null or valid uint160, at the moment you're allowing some arrays here, for example.

File: pkg/core/native/notary.go [hunk 42]
```diff
@@ -0,0 +1,358 @@
...
+		to, err = util.Uint160DecodeBytesBE(bytes)
+		if err != nil {
+			panic(fmt.Errorf("failed to decode `to` from `data` parameter: %w", err))
+		}
+	}
+	bigInt, err := additionalParams[1].TryInteger()
+	if err != nil {
+		panic(fmt.Errorf("failed to decode `till` from `data` parameter: %w", err))
+	}
+	till := uint32(bigInt.Int64())
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:172]: Better check it first for `IsInt64()` and then for fitting into `uint32` as it's invalid otherwise.

File: pkg/core/native/notary.go [hunk 43]
```diff
@@ -0,0 +1,358 @@
...
+func (n *Notary) lockDepositUntil(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	addr := toUint160(args[0])
+	ok, err := runtime.CheckHashedWitness(ic, addr)
+	if err != nil {
+		panic(fmt.Errorf("failed to check witness for %s: %w", addr.StringBE(), err))
+	}
+	if !ok {
+		return stackitem.NewBool(false)
+	}
+	till := uint32(toBigInt(args[1]).Int64())
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:210]: Same thing here.

File: pkg/core/native/notary.go [hunk 44]
```diff
@@ -0,0 +1,358 @@
...
+	}
+	return stackitem.NewBigInteger(deposit.Amount)
+}
+
+// expirationOf Returns deposit lock height for specified address.
+func (n *Notary) expirationOf(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	acc := toUint160(args[0])
+	deposit := n.getDepositFor(ic.DAO, acc)
+	if deposit == nil {
+		panic(fmt.Errorf("deposit for %s not found", acc.StringBE()))
```
Comment (roman-khimov) [inline pkg/core/native/notary.go:285]: BTW, we should make this symmetric to `balanceOf`, just return 0.

**Review by roman-khimov — CHANGES_REQUESTED** [review]
Can we bring diff coverage to 80+? It's at 73 now, pretty close but can be improved.

File: pkg/core/helper_test.go (hunk 37 above)
Comment (fyrchik) [reply pkg/core/helper_test.go:419]: Also, does it work only for native?

Comment (AnnaShaleva) [reply pkg/core/helper_test.go]: Changed the name of the method and replaced some common code of `core` tests.

**Review by roman-khimov — APPROVED** [review]
